src/comparator/wrappers.py

```diff
@@ -355,24 +389,69 @@ def nested_messages(self) -> Dict[str, "Message"]:
                 self.path + (3, i,),
                 self.resource_database,
             )
+            # Exclude the auto-generated map_entries message, since
+            # the generated message does not have real source code location.
+            # Including those messages in the comparator will fail the source code
+            # information extraction.
             for i, message in enumerate(self.message_pb.nested_type)
+            if not message.options.map_entry
         }
+        # fmt: on
+
+    @property
+    def map_entries(self) -> Dict[str, Dict[str, Field]]:
+        # If the nested message is auto-generated map entry for the maps field,
+        # the message name is fieldName + 'Entry', and it has two nested fields (key, value).
+        #
+        # For maps fields:
+        # map<KeyType, ValueType> map_field = 1;
+        # The parsed descriptor looks like:
+        #    message MapFieldEntry {
+        #        option map_entry = true;
+        #        optional KeyType key = 1;
+        #        optional ValueType value = 2;
+        #    }
+        #    repeated MapFieldEntry map_field = 1;
+        map_entries = {}
+        for message in self.message_pb.nested_type:
+            if message.options.map_entry:
+                fields = {field.name: field for field in message.field}
+                if not fields["key"] or not fields["value"]:
+                    raise TypeError(
+                        "The auto-generated map entry message should have key and value fields."
+                    )
```

Comment:
I think we don't just want to check that the "key" and "value" fields map to truthy elements, we want to make sure that the fields are there.
```python
if not {"key", "value"} <= fields.keys() 
```

Reply:
Thanks! Just learned that we can use `<=` to check if whether the set on the left is a subset of the set on the right.

But I'm wondering how it differs from ` if not fields["key"] or not fields["value"]:` since it also checks if the fields `key` and `value` are there in the dict (`<=` is cleaner though).

Reply:
I see the point! So it might be chances that the `fields` map is empty, the subset method would avoid the keyError. Thanks!

Reply:
Exactly. Looking up a key that isn't there generates an error that is different from the one we want to generate.